### PR TITLE
Introduce a moose_map_find to get info which object failed the search in a map

### PR DESCRIPTION
**Pull request by GiudGiud**
I would have namespaced it under Moose but right now it conflicts with libmesh's map_find
if we have a better name  I can try again to name space it but map_find is pretty pretty good

### CODE REVIEW

File: framework/include/utils/MooseUtils.h [hunk 1]
```diff
@@ -1197,6 +1198,129 @@ isDigits(const std::string & str)
 {
   return std::all_of(str.begin(), str.end(), [](unsigned char c) { return std::isdigit(c); });
 }
+
+/**
```
Comment (GiudGiud) [inline framework/include/utils/MooseUtils.h:1202]: should I credit the copy-pasting from libmesh::Utility?

File: framework/include/base/Moose.h [hunk 2]
```diff
@@ -279,3 +279,12 @@ class ScopedCommSwapper
...
 } // namespace Moose
+
+// TODO namespace this inside MOOSE as Moose::map_find once libmesh namespace is separate
+// This macro is useful to get around the deficiencies of the C++ map 'at' and 'find' operators
+// at errors with an exception which requires catching to output nicely
+// find is a very verbose syntax
+// This should only be used in a class where name() and type() are defined. If not, rely on
+// libmesh_map_find instead
+#define moose_map_find(map, key)                                                                   \
+  MooseUtils::map_find((map), (key), name(), type(), __FILE__, __LINE__)
```
Comment (dschwen) [inline framework/include/base/Moose.h:290]: We can probably do that not as a macro but as a templated function with SFINAE   (passing in `this` - if you want you can wrap that in a macro). Then we can use it anywhere we like and get name/type if it's available, and fall back to vanilla libmesh_find if not.

Comment (GiudGiud) [reply framework/include/base/Moose.h]: Alright gave it a go.

I cant wait to try C++20 for this, SFINAE is heinous to work with

File: framework/include/utils/MooseUtils.h [hunk 3]
```diff
@@ -1197,6 +1198,96 @@ isDigits(const std::string & str)
...
+ * This function should not be called directly (although it can be),
+ * instead see the moose::map_find() macro.
```
Comment (dschwen) [inline framework/include/utils/MooseUtils.h:1204]:
```suggestion
 * instead see the moose_map_find() macro.
```

File: framework/include/utils/MooseUtils.h [hunk 4]
```diff
@@ -1197,6 +1198,96 @@ isDigits(const std::string & str)
...
+               line_number);
+  return it->second;
+}
+
+/**
+ * Version where this is being used outside of a class so `this` does not exist
+ * First template for non-const returns
+ * Second template for const returns
+ * We enforce that the const-ness of the Map matches the constness of the return argument
+ */
```
Comment (dschwen) [inline framework/include/utils/MooseUtils.h:1267]: I don't get this. Who calls this?

File: framework/include/utils/MooseUtils.h [hunk 5]
```diff
@@ -1197,6 +1198,96 @@ isDigits(const std::string & str)
...
+               line_number);
+  return it->second;
+}
+template <
+    typename Map,
+    typename Key,
+    typename T,
+    typename T::type,
+    typename T::name,
+    typename std::enable_if<!libMesh::Utility::is_streamable<Key>::value, Key>::type * = nullptr>
```
Comment (dschwen) [inline framework/include/utils/MooseUtils.h:1245]: So why check if the key is streamable if you never steam it anywhere anyways?

File: framework/include/utils/MooseUtils.h (hunk 4 above)
Comment (GiudGiud) [reply framework/include/utils/MooseUtils.h:1267]: this is called by the moose_map_find macro if the above substitution failed

File: framework/include/utils/MooseUtils.h (hunk 5 above)
Comment (GiudGiud) [reply framework/include/utils/MooseUtils.h:1245]: maybe I forgot to stream it but it should be

Comment (dschwen) [reply framework/include/utils/MooseUtils.h:1247]: It looks like this template is only enable if the key is _not_ streamable

File: framework/include/utils/MooseUtils.h (hunk 4 above)
Comment (dschwen) [reply framework/include/utils/MooseUtils.h:1271]: ok, after your push it makes lightly more sense, but you have the enable ifs backwards I think

Comment (dschwen) [reply framework/include/utils/MooseUtils.h:1271]: Nonetheless, if `this` does not exist the macro substitution will give you a non-existing symbol right away.

Comment (GiudGiud) [reply framework/include/utils/MooseUtils.h:1271]: yeah. Not much I can do about this?

**Comment by dschwen** [comment]
What do you think about making it a member of `MooseObject`? And then have a separate `Moose::findMap(...)` function for use outside of MooseObjects?

**Comment by GiudGiud** [comment]
MooseObject and Action ?
That's how I had it initially.

**Comment by dschwen** [comment]
> It s going to be a lot of definitions of the same thing
No, you could have all call out to the same helper function in the namespace. Just make the object pointer `nullptr` when you want top avoid calling out to name() / type().

**Comment by dschwen** [comment]
You could also introduce a common interface that both MooseObject and Action inherit from and pass a pointer to that.

**Comment by dschwen** [comment]
See https://github.com/libMesh/libmesh/pull/3680

**Comment by GiudGiud** [comment]
Considering my other PRs trying to make the code base clearer, this is going in the wrong direction.

**Pull request closed without merging**